account_fiscal_product_rule/models/product.py

```
@@ -0,0 +1,45 @@
+# Copyright 2022 Akretion France (http://www.akretion.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductCategory(models.Model):
+    _inherit = "product.category"
+
+    fiscal_position_product_rule_ids = fields.Many2many(
+        "account.fiscal.position.product.rule", string="Fiscal Rule"
+    )
+
+    def get_matching_product_fiscal_rule(self, fiscal_pos):
+        self.ensure_one()
+        return self.fiscal_position_product_rule_ids.filtered(
+            lambda r: r.fiscal_position_id == fiscal_pos
+        ) or (
+            self.parent_id
+            and self.parent_id.get_matching_product_fiscal_rule(fiscal_pos)
+        )
+
```

We should add a python constraint to be sure to not put 2 account.fiscal.position.rule related to the same fiscal position

something like

```
def check_no_duplicate_fiscal_position(self):
    for record in self:
        fps = []
        for rule in record.fiscal_position_product_rule_ids:
            if rule.fiscal_position_id in fps:
                raise ValidationError(...)
            else:
                fps.append(rule.fiscal_position_id)
```

To avoid duplicated code for this constraint. You can create a mixin (fiscal.position.product.rule.owner) and put the code for the field definition "fiscal_position_product_rule_ids" and the constrainte

---

+ add the test that add 2 fiscal rule with the same fiscal position, it should raise an error (you can test only the product or the categ as the code is the same).